File: example/http4s/consumer/src/main/scala/poppet/example/http4s/consumer/controller/UserController.scala
package poppet.example.http4s.consumer.controller

import cats.effect.IO
import io.circe.generic.auto._
import org.http4s._
import org.http4s.circe.CirceEntityCodec._
import org.http4s.dsl.io._
import poppet.example.http4s.consumer.service.UserServiceProvider

class UserController(userServiceProvider: UserServiceProvider) {
    private val userService = userServiceProvider.get

    val routes = HttpRoutes.of[IO] {
        case GET -> Root / "api" / "user" / id =>
            userService.findById(id).foldF(InternalServerError(_), Ok(_))
    }
}

File: example/spring/consumer/src/poppet/example/spring/controller/UserController.java
package poppet.example.spring.controller;

import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import poppet.example.spring.model.User;
import poppet.example.spring.service.UserService;

@RestController
public class UserController {
    private final UserService userService;

    public UserController(UserService userService) {
        this.userService = userService;
    }

    @RequestMapping("/api/user/{id}")
    public User findById(@PathVariable("id") String id) {
        return userService.findById(id);
    }
}
